src/spacetime.jl

```diff
@@ -106,6 +106,208 @@ end
 """
 default_ntmod(dtrec::Real, dtmod::Real, ntrec::Int) = default_ntmod(dtrec, dtmod, 0.0, ntrec)[2]
 
+
+
+#
+# time shift codes
+#
+mutable struct TimeShift{T<:Real}
+    h::Vector{T}
+    nshift::Int
+    shift_fraction::T
+    bc::AbstractString
+end
+
+"""
+    h = WaveFD.shiftfilter(shift; [length=12, α=1.0, bc="zero"])
+
+Build a cosine tapered sinc filter for shifting an array by a decimal number of samples. The optional parameters are:
+
+* `length::Int` length of the filter (better to be an even number)
+* `α::Real` parameter to control the strength of the cosine taper. α=0 -> no taper
+* `bc::String` how to deal with the edges. Currently the two options are "zero" and "nearest".
+
+# Notes
+It is checked that `0 <= α <= 1`.
+The taper is centered to the middle of the filter and goes all the way to the edges.
+For α = 1, the value of the taper at the edges is ≈ 0.
+The shift is split into integer number of samples and a residual fractional sample.
+"""
+function shiftfilter(shift::T; length::Int=12, α::Real=1.0, bc::AbstractString="zero") where T<:Real
+
+    @assert 0 <= α <= 1
+    @assert bc ∈ ["zero","nearest"]
+    nshift = floor(shift)
+    shift_fraction = shift - nshift
+
+    hl = div(length,2) + length % 2
+    j = [1:length;] .- (div(length,2) + 1 )
+    a = sinc.(j .+ shift_fraction) .* ((1-α) .+ α/2*(1 .+ cos.(π*(j .+ shift_fraction)/hl)))
+    a = a./sum(a)
+
+    return TimeShift{T}(a, nshift, shift_fraction, bc)
+end
+
+"""
+    WaveFD.shiftforward!(h, d, m, bc)
+
+Shift `m::Array{T,N}` to `d::Array{T,N}` by a decimal number of samples. `d` and `m` have the same size.
+A positive shift will delay the trace. `h` is built using `WaveFD.shiftfilter`. Both filter and amount 
+of shifting must be shorter than the arrays, otherwise, an assertion will fail. 
+For examples:
+
+    WaveFD.shiftforward!(WaveFD.shiftfilter(-3.3), d, m)
+    WaveFD.shiftforward!(WaveFD.shiftfilter(13.34, length=12, α=1.0, bc="zero"), d, m)
+
+Note that `N=1`, `N=2` or `N=3` are supported. If `N=2` or `N=3`, then interpolation is done along the fast dimension.
+"""
+function shiftforward!(H::TimeShift{T0}, d::StridedArray{T,1}, m::StridedArray{T,1}) where {T0<:Real, T<:Real}
+    
+    @assert length(m) == length(d)
+
+    if H.nshift==0 && H.shift_fraction ≈ 0
+        d .= m
+        return nothing
+    end
+
+    n = length(m)
+    nshift = H.nshift
+    nfilter = length(H.h)
+    n_half_filter = div(nfilter,2) + nfilter % 2
+    bc_multiplier = (H.bc=="nearest") ? 1 : 0
+
+    @assert n > n_half_filter
+    @assert nshift < n 
+
+    # create a padded array
+    mpad = Vector{T}(undef, n + 2*n_half_filter)
+    mpad[1+n_half_filter:n+n_half_filter] = m[:]
```

Comment:
```suggestion
    mpad[1+n_half_filter:n+n_half_filter] .= m
```
I might be wrong, but I think that `m[:]` might create a copy.